Source/JavaScriptCore/runtime/Completion.cpp

```diff
@@ -266,4 +266,56 @@ JSInternalPromise* importModule(JSGlobalObject* globalObject, const Identifier&
     return globalObject->moduleLoader()->requestImportModule(globalObject, moduleKey, parameters, scriptFetcher);
 }
 
+HashMap<RefPtr<UniquedStringImpl>, String> retrieveAssertionsFromDynamicImportOptions(JSGlobalObject* globalObject, JSValue options, const HashSet<RefPtr<UniquedStringImpl>>& supportedAssertions)
+{
+    // https://tc39.es/proposal-import-assertions/#sec-evaluate-import-call
+
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (options.isUndefined())
+        return { };
+
+    auto* optionsObject = jsDynamicCast<JSObject*>(options);
+    if (UNLIKELY(!optionsObject)) {
+        throwTypeError(globalObject, scope, "dynamic import's options should be an object"_s);
+        return { };
+    }
+
+    JSValue assertions = optionsObject->get(globalObject, vm.propertyNames->builtinNames().assertPublicName());
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (assertions.isUndefined())
+        return { };
+
+    auto* assertionsObject = jsDynamicCast<JSObject*>(assertions);
+    if (UNLIKELY(!assertionsObject)) {
+        throwTypeError(globalObject, scope, "dynamic import's options.assert should be an object"_s);
+        return { };
+    }
+
+    PropertyNameArray properties(vm, PropertyNameMode::Strings, PrivateSymbolMode::Exclude);
+    assertionsObject->methodTable()->getOwnPropertyNames(assertionsObject, globalObject, properties, DontEnumPropertiesMode::Exclude);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    HashMap<RefPtr<UniquedStringImpl>, String> result;
+    for (auto& key : properties) {
+        JSValue value = assertionsObject->get(globalObject, key);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (UNLIKELY(!value.isString())) {
+            throwTypeError(globalObject, scope, "dynamic import's options.assert includes non string property"_s);
+            return { };
+        }
+
+        String valueString = value.toWTFString(globalObject);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (supportedAssertions.contains(key.impl()))
```

Comment:
Using HashSet for this especially when there is exactly one item seems wasteful to me.
Why don't we just have a Vector instead?
It's probably a lot quicker to go through Vector of 3-4 items than going through a hashing.

Reply:
OK, changed it to Vector.